Address generator unittest

Ceazer101 opened the pull request.

Commit:
address generator functionality

Commit:
trying dependencies

Commit:
Merge branch 'main' of https://github.com/KeaTesting/PersonalTestDataGeneratorBackend into AddressGenerator-unittest

Commit:
setup branch changes

Commit:
commit staged again

Commit:
deleted duplicated file

Commit:
changed methods from private to public

Commit:
made unit tests for methods

Ceazer101 committed:
Merge branch 'main' into AddressGenerator-unittest

victor-dat21b commented on PersonalTestDataGeneratorBackend/PersonalTestDataGeneratorBackend/AddressGenerator.cs:
```diff
@@ -0,0 +1,65 @@
+﻿namespace PersonalTestDataGeneratorBackend.PersonalTestDataGeneratorBackend
```
REJECTED

Bjand49 replied:
Lidt som victor siger, men med en lidt mere konstruktiv besked😅

Du må gerne tage filerne ud af mappen, og sørge for at opdatere namespacet. Jeg opdaterede det i det pull request der kom ind tidligere, men det er med her. 
"PersonalTestDataGeneratorBackend.PersonalTestDataGeneratorBackend" giver ikke så meget mening

Troels-dat21B replied:
based

victor-dat21b commented on PersonalTestDataGeneratorBackend/PersonalTestDataGeneratorBackend/AddressGenerator.cs:
```diff
@@ -0,0 +1,65 @@
+            return letter.HasValue ? $"{number}{letter}" : number.ToString();
+        }
+
+        //Floor. Either “st” or a number from 1 to 99
```
bad comment cringe

Ceazer101 committed:
Merge branch 'main' into AddressGenerator-unittest

Bjand49 commented on PersonalTestDataGeneratorBackend/PersonalTestDataGeneratorBackend/addresses.sql:
```diff
@@ -0,0 +1,667 @@
+-- phpMyAdmin SQL Dump
```
Det her må gerne komme ind i en migration

Commit:
moved file and changed namespace

Troels-dat21B commented on PersonalTestDataGeneratorBackend/PersonalTestDataGeneratorBackend/AddressGenerator.cs:
```diff
@@ -0,0 +1,65 @@
+        public string GenerateNumber()
+        {
+            int number = random.Next(1, 1000);
+            char? letter = random.Next(0, 2) == 1 ? (char?)random.Next('A', 'Z' + 1) : null;
```
Hvorfor er max nummeret 100, når opgaven siger op til 999? :)

Troels-dat21B commented on PersonalTestDataGeneratorBackend/PersonalTestDataGeneratorBackend/AddressGenerator.cs:
```diff
@@ -0,0 +1,65 @@
+        //Postal code and town. Randomly extracted from the provided database addresses.sql
+        public static string? GeneratePostalCode()
+        {
+            //implement this, when the database is made and has the data from the addresses.sql
```
Er denne glemt? Eller er det har, du hentydede til at dette ikke var færdigt?

Ceazer101 replied:
den er lavet nu

Bjand49 commented on PersonalTestDataGeneratorBackend/PersonalTestDataGeneratorBackend/AddressGenerator.cs:
```diff
@@ -0,0 +1,65 @@
+        public string GenerateNumber()
+        {
+            int number = random.Next(1, 1000);
+            char? letter = random.Next(0, 2) == 1 ? (char?)random.Next('A', 'Z' + 1) : null;
```
Hvorfor kan letter være null? Giver det ikke mening at man bare giver den en tom værdi så den ikke skal valideres for null senere hen?

Ceazer101 replied:
fordi det er en 50% chance for om letter kommer med, så hvis der ikke er letter bliver den nød til at være null

Bjand49 replied:
Der er ingen grund til at der skal returners null her. du kan skrive '' til sidst for at sikre at der er en værdi der ikke er null

Ceazer101 replied:
men den retunere ikke null den retunere enten en værdi med bare tal eller tal og bogstav

Bjand49 commented on PersonalTestDataGeneratorBackend/PersonalTestDataGeneratorBackend/AddressGenerator.cs:
```diff
@@ -0,0 +1,65 @@
+        public static string? GeneratePostalCode()
+        {
+            //implement this, when the database is made and has the data from the addresses.sql
+            return null;
```
Skal implementeres før den kan komme ind

Troels-dat21B replied:
That's what I said!

Commit:
deleted adresses.sql and emprty folder

Commit:
corrected namespace

Commit:
added PostalCode entity

Commit:
added PostalCode DB relation

Commit:
migrateted PostalCode with all the adress data

Commit:
Merge branch 'main' of https://github.com/KeaTesting/PersonalTestDataGeneratorBackend into AddressGenerator-unittest

Commit:
added address to classes: Person and Personhelper

Commit:
added migrations with address added to person

Commit:
separated address responsibility into service and repo and added repo so PostalCodes can be recieved from DB

Commit:
added unit test and mock for PostalCode method

Commit:
added connection to db

Ceazer101 committed:
Merge branch 'main' into AddressGenerator-unittest

Bjand49 commented on PersonalTestDataGeneratorBackend/Generators/AddressGenerator.cs:
```diff
@@ -0,0 +1,65 @@
+        public static string? GeneratePostalCode()
+        {
+            //implement this, when the database is made and has the data from the addresses.sql
+            return null;
```
Var det her ikke lavet?

Bjand49 commented on UnitTests/AddressGeneratorUnitTest.cs:
```diff
@@ -0,0 +1,88 @@
+            string number = _addressGenerator.GenerateNumber();
+
+            // Assert
+            Assert.Matches(@"^\d{1,1000}[A-Z]?$", number);
```
Skal det ikke være muligt at kunne returnere et vejnavn med et mellemrum i?

Ceazer101 replied:
hans beskrivelse
A random assortment of alphabetic characters

Bjand49 commented:
Der mangler alt det vi sad med igår, gør der ikke?

Commit:
Merge branch 'AddressGenerator-unittest' of https://github.com/KeaTesting/PersonalTestDataGeneratorBackend into AddressGenerator-unittest

Bjand49 commented on PersonalTestDataGeneratorBackend/DB/GeneratorDB.cs:
```diff
@@ -8,5 +8,14 @@ public GeneratorDB(DbContextOptions options) : base(options) { }
+        {
+            if (!optionsBuilder.IsConfigured)
+            {
+                optionsBuilder.UseMySql("Server=localhost;Port=3307;Database=mydatabase;Uid=root;Pwd=12345;", ServerVersion.AutoDetect("Server=localhost;Port=3307;Database=mydatabase;Uid=root;Pwd=12345;"));
```
Vi snakkede om det, prøv at se om du kan få den her værdi ud af koden og over i et centralt sted så der ikke er duplicates

Bjand49 replied:
Sonar cloud fejler også på den, så den skal ud

Bjand49 commented on PersonalTestDataGeneratorBackend/Generators/AddressGenerator.cs:
```diff
@@ -0,0 +1,82 @@
+        public string GenerateStreet()
+        {
+            int length = random.Next(5, 15);
+            const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
```
Vores gader kan også have æøå samt mellemrum i

Bjand49 commented on PersonalTestDataGeneratorBackend/Generators/AddressGenerator.cs:
```diff
@@ -0,0 +1,82 @@
+        public string GenerateNumber()
+        {
+            int number = random.Next(1, 1000);
+            char? letter = random.Next(0, 2) == 1 ? (char?)random.Next('A', 'Z' + 1) : null;
```
Der er ingen grund til at der skal returners null her. du kan skrive '' til sidst for at sikre at der er en værdi der ikke er null

Bjand49 commented on UnitTests/AddressGeneratorUnitTest.cs:
```diff
@@ -0,0 +1,127 @@
+        }
+
+        [Fact]
+        public void GenerateFloor_ShouldReturnStSometimes()
```
Det her er ikke en testcase der giver mening i dens nuværende tilstand. Den er dækket ind af den test der er ovenfor der hedder GenerateFloor_ShouldReturnStOrNumber. Hvis du eksplicit vil teste om er er en "st" med, så se om du kan gøre det på en ikke random måde

Bjand49 commented on UnitTests/AddressGeneratorUnitTest.cs:
```diff
@@ -0,0 +1,127 @@
+        public void GenerateFloor_ShouldReturnNumberBetween1And29()
+        {
+            // Act
+            var results = Enumerable.Range(0, 100).Select(_ => _addressGenerator.GenerateFloor()).Where(floor => floor != "st").ToList();
```
Den her test kan potentielt fejle fordi vi regner med at der er mindst en af de 100 der har et nummer i sig. Det er ikke godt at være afhængig af at bruge en random når vi tester.

Bjand49 replied:
Prøv at se om du kan finde på en anden måde at gøre det på, uden at bruge random

Bjand49 commented on UnitTests/AddressGeneratorUnitTest.cs:
```diff
@@ -0,0 +1,127 @@
+            string result = _addressGenerator.GenerateDoor();
+
+            // Assert
+            Assert.True(
```
Der sker mange tests i den her, overvej at se om du kan dele det op i flere små tests

Commit:
added and implemented Randomgenerator with Interface to mock random values

Commit:
implemented RandomGenerator class where it was needed,
added æøå to methods,
removed null from GenerateNumber(),
corrected range from 0, 1000 to 1, 1000

Commit:
added mocking of random values to unit tests part1 will continue on friday

Commit:
Now db can be read from appsettings.json

Commit:
Merge branch 'main' of https://github.com/KeaTesting/PersonalTestDataGeneratorBackend into AddressGenerator-unittest

Bjand49 committed:
Removed randomgenerator and refactored functionality to be testable

Bjand49 committed:
removed duplicate tests

Bjand49 committed:
refactored floor generation

Bjand49 committed:
updated tests

Bjand49 committed:
Fixed test final

victor-dat21b reviewed (approved):
Cleaned up nicely, good job
